src/Microsoft.OpenApi/Services/OpenApiUrlSpaceNode.cs

```diff
@@ -0,0 +1,231 @@
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.OpenApi.Extensions;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.OpenApi.Services
+{
+    /// <summary>
+    /// A directory structure representing the paths of an OpenAPI document.
+    /// </summary>
+    public class OpenApiUrlSpaceNode
+    {
+        /// <summary>
+        /// All the subdirectories of a node.
+        /// </summary>
+        public IDictionary<string, OpenApiUrlSpaceNode> Children { get; set; } = new Dictionary<string, OpenApiUrlSpaceNode>();
+
+        /// <summary>
+        /// The name tag for a group of nodes.
+        /// </summary>
+        public string Label { get; set; }
+
+        /// <summary>
+        /// Path Item object that describes the operations available on a node.
+        /// </summary>
+        public OpenApiPathItem PathItem { get; private set; }
+
+        /// <summary>
+        /// The relative directory path of the current node from the root node.
+        /// </summary>
+        public string Path { get; set; } = "";
+
+        /// <summary>
+        /// Flag indicating whether a node segment is a path parameter.
+        /// </summary>
+        public bool IsParameter => Segment.StartsWith("{");
+
+        /// <summary>
+        /// Flag indicating whether a node segment is a path function.
+        /// </summary>
+        public bool IsFunction => Segment.Contains("(");
+
+        /// <summary>
+        /// The subdirectory of a relative path.
+        /// </summary>
+        public string Segment { get; private set; }
+
+        /// <summary>
+        /// The Pascal-cased alphabet name of a segment.
+        /// </summary>
+        public string Identifier
+        {
+            get
+            {
+                string identifier;
+                if (IsParameter)
+                {
+                    identifier = Segment.Substring(1, Segment.Length - 2).ToPascalCase();
+                }
+                else
+                {
+                    identifier = Segment.ToPascalCase().Replace("()", "");
+                    var openParen = identifier.IndexOf("(");
+                    if (openParen >= 0)
+                    {
+                        identifier = identifier.Substring(0, openParen);
+                    }
+                }
+                return identifier;
+            }
+        }
+
+        /// <summary>
+        /// Flag indicating whether a PathItem has operations.
+        /// </summary>
+        /// <returns>true or false.</returns>
+        public bool HasOperations()
+        {
+            return PathItem != null && PathItem.Operations != null && PathItem.Operations.Count > 0;
+        }
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="segment">The subdirectory of a relative path.</param>
+        public OpenApiUrlSpaceNode(string segment)
```

Comment:
should this be private since there's a create method

Reply:
I agree. 
@darrelmiller what was the use case for making this public? I can only see it's being called outside the class in one of Kiota's tests [here](https://github.com/microsoft/kiota/blob/60e831d24c8cdc8cbdc183a9d9f7a9a1343ac889/tests/kiota.core.tests/KiotaBuilderTests.cs#L12) but nowhere else.
We could instead update the `Create()` [method](https://github.com/irvinesunday/OpenAPI.NET/blob/876611bfa610893fb50b413321cb86b3f2380fe0/src/Microsoft.OpenApi/Services/OpenApiUrlSpaceNode.cs#L142) for clients to use when they want to create a 'blank' `OpenApiUrlSpace` node as below: 
```
 public static OpenApiUrlSpaceNode Create(OpenApiDocument doc = null, string label = string.Empty)
        {
            OpenApiUrlSpaceNode root = new OpenApiUrlSpaceNode(string.Empty);

            var paths = doc?.Paths;
            if (paths != null)
            {
                root = new OpenApiUrlSpaceNode(string.Empty);

                foreach (var path in paths)
                {
                    root.Attach(path.Key, path.Value, label);
                }
            }
            return root;
        }
``` 
```
var node = OpenApiUrlSpaceNode.Create();
```

Reply:
I'm fine with making that internal/private

Reply:
Yes, a create method with no parameters would also be useful if we wanted to create a tree from scratch